---
title: "Assessing orpheum's efficacy for ORF prediction from prokaryotic metagenomes"
output: 
  pdf_document:
    fig_caption: yes
    includes:  
      in_header: knit.tex
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, include = T, eval = T, warning = F,
                      cache = T)
options(scipen=999)
```

```{r libraries}
library(dplyr)
library(readr)
library(tibble)
library(tidyr)
library(ggplot2)
library(ggpmisc)
library(ggpubr)
library(broom)
library(kableExtra)
library(rjson)
library(purrr)
library(janitor)
```

```{r ggplot_base}
ggplot_base <- theme_minimal() +
  theme(plot.title = element_text(size = 8),
        axis.text = element_text(size = 6),
        axis.title = element_text(size = 7),
        legend.title = element_text(size = 6), 
        legend.text = element_text(size = 6),
        plot.title.position = "plot")
  #theme(plot.title.position = "plot")
```

```{r function_read_and_format_orpheum_json}
read_and_format_orpheum_json <- function(sys_glob_path, database, alphabet, ksize) {
  file_prefix = paste0("outputs\\/orpheum\\/", database, "\\/", alphabet, "_ksize", ksize, "\\/")
  json_df <- Sys.glob(sys_glob_path) %>%
    set_names() %>%
    map_dfr(~RJSONIO::fromJSON(content =., nullValue = NaN)$categorization_counts, .id = "sample") %>%
    clean_names() %>%
    mutate(sample = gsub(file_prefix, "", sample)) %>%
    mutate(sample = gsub("_GCF_900036035\\.1_RGNV35913_genomic\\.fna\\.gz\\.cdbg_ids\\.reads\\.summary\\.json", "", sample)) %>%
    mutate(database = database) %>%
    mutate(ksize = paste0("k=", ksize)) %>%
    mutate(alphabet = alphabet) %>%
    mutate(nbhd_reads = rowSums(across(where(is.numeric))))
     
  return(json_df)
}

read_and_format_orpheum_json_translation <- function(sys_glob_path, database, alphabet, ksize) {
  file_prefix = paste0("outputs\\/orpheum\\/", database, "\\/", alphabet, "_ksize", ksize, "\\/")
  json_df <- Sys.glob(sys_glob_path) %>%
    set_names() %>%
    map_dfr(~RJSONIO::fromJSON(content =., nullValue = NaN)$histogram_n_coding_frames_per_read, .id = "sample") %>%
    clean_names() %>%
    mutate(sample = gsub(file_prefix, "", sample)) %>%
    mutate(sample = gsub("_GCF_900036035\\.1_RGNV35913_genomic\\.fna\\.gz\\.cdbg_ids\\.reads\\.summary\\.json", "", sample)) %>%
    mutate(database = database) %>%
    mutate(ksize = paste0("k=", ksize)) %>%
    mutate(alphabet = alphabet)
  
  # make sure numbers are encoded as numeric, then replace NAs with 0s and sum to coding
  json_df <- json_df %>%
    mutate(across(where(~ anyNA(.) & is.numeric(.)), ~ replace_na(., 0))) %>%
    mutate(coding_reads = rowSums(across(where(is.numeric))))
  
  # add missing cols
  columns <- c(number_of_reads_with_1_putative_protein_coding_translations = 0,
               number_of_reads_with_2_putative_protein_coding_translations = 0,
               number_of_reads_with_3_putative_protein_coding_translations = 0,
               number_of_reads_with_4_putative_protein_coding_translations = 0,
               number_of_reads_with_5_putative_protein_coding_translations = 0,
               number_of_reads_with_6_putative_protein_coding_translations = 0)
  json_df <- json_df %>%
    add_column(!!!columns[!names(columns) %in% names(.)])
  
  # calculate percentage of reads translated into n number of coding frames
  json_df <- json_df %>%
    mutate(f_putative_coding_1_translation = ifelse(number_of_reads_with_1_putative_protein_coding_translations == 0, 0,
                                                    number_of_reads_with_1_putative_protein_coding_translations/coding_reads)) %>%
    mutate(f_putative_coding_2_translation = ifelse(number_of_reads_with_2_putative_protein_coding_translations == 0, 0,
                                                    number_of_reads_with_2_putative_protein_coding_translations/coding_reads)) %>%
    mutate(f_putative_coding_3_translation = ifelse(number_of_reads_with_3_putative_protein_coding_translations == 0, 0,
                                                    number_of_reads_with_3_putative_protein_coding_translations/coding_reads)) %>%
    mutate(f_putative_coding_4_translation = ifelse(number_of_reads_with_4_putative_protein_coding_translations == 0, 0,
                                                    number_of_reads_with_4_putative_protein_coding_translations/coding_reads)) %>%
    mutate(f_putative_coding_5_translation = ifelse(number_of_reads_with_5_putative_protein_coding_translations == 0, 0,
                                                    number_of_reads_with_5_putative_protein_coding_translations/coding_reads)) %>%
    mutate(f_putative_coding_6_translation = ifelse(number_of_reads_with_6_putative_protein_coding_translations == 0, 0,
                                                    number_of_reads_with_6_putative_protein_coding_translations/coding_reads))
  
  return(json_df)
}
```

```{r function_plot_orpheum}
plot_orpheum_translation_frames <- function(orpheum_trans, ksize, alphabet, database){
  orpheum_trans_long <- orpheum_trans %>%
    select(sample, database, ksize, alphabet, starts_with("f_putative")) %>%
    pivot_longer(cols = starts_with("f_putative"), names_to = "translation_frame", values_to = "fraction") %>%
    mutate(translation_frame = gsub("f_putative_coding_", "", translation_frame)) %>%
    mutate(translation_frame = gsub("_translation", "", translation_frame))
  
  ggplot(orpheum_trans_long, aes(x = translation_frame, y = fraction)) +
    geom_violin(scale = "width") +
    theme_minimal() +
    ggplot_base +
    geom_segment(aes(x = .5, y = .91, xend = 1.5, yend = .91), linetype = "dashed") +
    geom_segment(aes(x = .5, y = .99, xend = 1.5, yend = .99), linetype = "dashed") +
    geom_segment(aes(x = 1.5, y = .09, xend = 2.5, yend = .09), linetype = "dashed") +
    geom_segment(aes(x = 1.5, y = .01, xend = 2.5, yend = .01), linetype = "dashed") +
    labs(x = "number of coding frames per read", y = "fraction of putative coding reads",
         title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize))
}
```

```{r function_plot_flagstat}
plot_flagstat_aa <- function(flagstat_path, nbhd_reads, ksize, alphabet, database){
  orph_aa_vs_aa_ref <- read_tsv(flagstat_path) %>%
    mutate(sample = gsub("_GCF_900036035.1_RGNV35913_genomic.fna", "", Sample)) %>%
    left_join(nbhd_reads) %>%
    mutate(original_pct = (total_passed/nbhd_reads)*100)
  
  mean_original_pct <- orph_aa_vs_aa_ref %>% 
    pull(original_pct) %>% 
    mean() %>%
    signif(3)
  
  original <- ggplot(orph_aa_vs_aa_ref, aes(x = original_pct)) +
    geom_density() +
    ggplot_base +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         x = "% original reads mapped to AA ref") +
    geom_vline(aes(xintercept = mean(original_pct)), col='grey', linetype = "dashed") +
    geom_text(aes(label=paste0("mean = ", mean_original_pct), 
                  x=mean_original_pct/2, y = Inf), size = 4, vjust = 1)
  
  mean_mapped_passed_pct <- orph_aa_vs_aa_ref %>% 
    pull(mapped_passed_pct) %>% 
    mean() %>%
    signif(3)
  
  total <- ggplot(orph_aa_vs_aa_ref, aes(x = mapped_passed_pct)) +
    geom_density() +
    ggplot_base +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         x = "% predicted reads mapped to AA ref") +
    geom_vline(aes(xintercept = mean(mapped_passed_pct)), col='grey', linetype = "dashed") +
    #geom_text(aes(x=mean_mapped_passed_pct/2, label=paste0("Mean\n", mean_mapped_passed_pct), y=.06))
    geom_text(aes(label=paste0("mean = ", mean_mapped_passed_pct), 
                  x=mean_mapped_passed_pct/2, y = Inf), size = 4, vjust = 1)

  return(ggarrange(original, total))
}

plot_predicted_aa_and_flagstat_aa <- function(orpheum_json, flagstat_path, ksize, alphabet, database, noncoding_estimate){

  orph_aa_vs_reference <- read_tsv(flagstat_path) %>%
    replace_na(list(mapped_passed_pct = 0)) %>%
    mutate(Sample = gsub("_GCF_900036035\\.1_RGNV35913_genomic\\.fna", "", Sample)) %>%
    left_join(orpheum_json, by = c("Sample" = "sample")) %>%
    mutate(coding_pct = (coding/nbhd_reads)*100)
  
  # format for geom_col
  tmp <- orph_aa_vs_reference %>%
    select(Sample, coding_pct, mapped_passed_pct) %>%
    mutate(coding_mapped_pct = coding_pct * (mapped_passed_pct/100)) %>%
    mutate(coding_unmapped_pct = coding_pct - coding_mapped_pct)  %>%
    select(-mapped_passed_pct) %>%
    pivot_longer(cols = ends_with("mapped_pct"), names_to = "fraction", values_to = "percent")
  
  plt <- ggplot(tmp, aes(x = reorder(Sample, coding_pct), y = percent, fill = fraction)) +
    geom_col() +
    geom_hline(aes(yintercept = 100 - noncoding_estimate), color = "black", linetype = "dashed", ) +
    ggplot_base +
    ylim(c(0, 100)) +
    theme(axis.text.x = element_blank(),
          legend.position = "bottom") +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         y = "putative coding reads\n(% of total library)",
         x = "genome query neighborhood", fill = "mapped to reference\namino acid CDS") +
    scale_fill_discrete(labels = c("mapped\n(true positive)", "unmapped (not\nin ref or false positive)"))
  plt  
  return(plt)
}

plot_flagstat_nnc <- function(flagstat_path, ksize, alphabet, database){
  orph_nnc_vs_aa_ref <- read_tsv(flagstat_path) %>%
    replace_na(list(mapped_passed_pct = 0))
  
  mean_mapped_passed_pct <- orph_nnc_vs_aa_ref %>% 
    pull(mapped_passed_pct) %>% 
    mean() %>%
    signif(3)
  
  plt <- ggplot(orph_nnc_vs_aa_ref, aes(x = mapped_passed_pct)) +
    geom_density() +
    ggplot_base +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         x = "% nucleotide noncoding reads mapped to nuc ref") +
    geom_vline(aes(xintercept = mean(mapped_passed_pct)), col='grey', linetype = "dashed") +
    geom_text(aes(label=paste0("mean = ", mean_mapped_passed_pct), 
                  x=mean_mapped_passed_pct/2, y = Inf), size = 4, vjust = 1)

  return(plt)
}

plot_flagstat_nc <- function(flagstat_path, ksize, alphabet, database){
  orph_nc_vs_aa_ref <- read_tsv(flagstat_path) %>%
    replace_na(list(mapped_passed_pct = 0))
  
  mean_mapped_passed_pct <- orph_nc_vs_aa_ref %>% 
    pull(mapped_passed_pct) %>% 
    mean() %>%
    signif(3)
  
  plt <- ggplot(orph_nc_vs_aa_ref, aes(x = mapped_passed_pct)) +
    geom_density() +
    ggplot_base +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         x = "% nucleotide coding reads mapped to nuc ref") +
    geom_vline(aes(xintercept = mean(mapped_passed_pct)), col='grey', linetype = "dashed") +
    geom_text(aes(label=paste0("mean = ", mean_mapped_passed_pct), 
                  x=mean_mapped_passed_pct/2, y = Inf), size = 4, vjust = 1)
  
  return(plt)
}

plot_predicted_nnc_vs_flagstat_nnc <- function(orpheum_json, flagstat_path, ksize, alphabet, database, noncoding_estimate){

  orph_nnc_vs_reference <- read_tsv(flagstat_path) %>%
    replace_na(list(mapped_passed_pct = 0)) %>%
    mutate(Sample = gsub("_GCF_900036035\\.1_RGNV35913_genomic\\.fna", "", Sample)) %>%
    left_join(orpheum_json, by = c("Sample" = "sample"))
  
  plt <- ggplot(orph_nnc_vs_reference, aes(x = (non_coding/nbhd_reads)*100, y = mapped_passed_pct)) +
    #geom_rect(aes(xmin = 3, xmax = 60, ymin=0, ymax=Inf), color = "#2C77BF", fill = alpha("#2C77BF", .1)) + 
    annotate(geom = "rect", xmin = 3, xmax = 60, ymin=0, ymax=Inf, fill = "#2C77BF", alpha = 0.2) +
    geom_vline(aes(xintercept = noncoding_estimate), color = "#2C77BF") +
    geom_point(alpha = 0.1) +
    ggplot_base +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         x = "putative non coding reads\n(% of total library)",
         y = "putative non coding reads mapped to reference\n(% of putative non coding reads)") +
    geom_hline(aes(yintercept = mean(mapped_passed_pct)), col='grey', linetype = "dashed") +
    geom_vline(aes(xintercept = mean((non_coding/nbhd_reads)*100)),  col='grey', linetype = "dashed")
    

  return(plt)
}

plot_predicted_nnc_and_flagstat_nnc <- function(orpheum_json, flagstat_path, ksize, alphabet, database, noncoding_estimate){

  orph_nnc_vs_reference <- read_tsv(flagstat_path) %>%
    replace_na(list(mapped_passed_pct = 0)) %>%
    mutate(Sample = gsub("_GCF_900036035\\.1_RGNV35913_genomic\\.fna", "", Sample)) %>%
    left_join(orpheum_json, by = c("Sample" = "sample")) %>%
    mutate(non_coding_pct = (non_coding/nbhd_reads)*100)
  
  # format for geom_col
  tmp <- orph_nnc_vs_reference %>%
    select(Sample, non_coding_pct, mapped_passed_pct) %>%
    mutate(non_coding_mapped_pct = non_coding_pct * (mapped_passed_pct/100)) %>%
    mutate(non_coding_unmapped_pct = non_coding_pct - non_coding_mapped_pct)  %>%
    select(-mapped_passed_pct) %>%
    pivot_longer(cols = ends_with("mapped_pct"), names_to = "fraction", values_to = "percent")
  
  plt <- ggplot(tmp, aes(x = reorder(Sample, non_coding_pct), y = percent,
                                           fill = fraction)) +
    geom_col() +
    geom_hline(aes(yintercept = noncoding_estimate), color = "black", linetype = "dashed", ) +
    ggplot_base +
    ylim(c(0, 100)) +
    theme(axis.text.x = element_blank(),
          legend.position = "bottom") +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize), 
         y = "putative non coding reads\n(% of total library)",
         x = "genome query neighborhood", fill = "mapped to reference\nnucleotide CDS") +
    scale_fill_discrete(labels = c("mapped\n(false negative)", "unmapped"),
                        guide = guide_legend(reverse = TRUE))
    
  return(plt)
}
```

```{r function_plot_stat}
plot_stat_nnc <- function(sys_glob_stat, ksize, alphabet, database){
  file_prefix <- strsplit(sys_glob_stat, split = "\\*")[[1]][1]
  orph_nnc_vs_aa_ref_stat <- Sys.glob(sys_glob_stat) %>%
    set_names() %>%
    map_dfr(read_tsv, col_names = c("stat", "value", "comment"), .id = "sample") %>%
    mutate(sample = gsub(file_prefix, "", sample)) %>%
    mutate(sample = gsub("_GCF_900036035.1_RGNV35913_genomic.fna.gz.cdbg_ids.reads.nuc_noncoding.stat", "", sample)) %>%
    mutate(stat = gsub(":", "", stat)) 
  
  stat_names_df <- data.frame(old_names = unique(orph_nnc_vs_aa_ref_stat$stat), 
                              new_names = unique(orph_nnc_vs_aa_ref_stat$stat))
  stat_names_df$new_names <- make_clean_names(stat_names_df$new_names)
  
  orph_nnc_vs_aa_ref_stat <- orph_nnc_vs_aa_ref_stat %>%
    left_join(stat_names_df, by = c("stat" = "old_names")) %>%
    select(-stat, stat = new_names)
  
  mean_error_rate <- orph_nnc_vs_aa_ref_stat %>% 
    filter(stat == "error_rate") %>%
    pull(value) %>% 
    mean() %>%
    signif(2)
  
  ggplot(orph_nnc_vs_aa_ref_stat %>% filter(stat == "error_rate"), aes(x = value)) +
    geom_density() +
    labs(title = paste0("db = ", database, ", alphabet = ", alphabet, ", k = ", ksize),  
         x = "mapping error rate") +
    ggplot_base +
    geom_vline(aes(xintercept = mean(value)), col='grey', linetype = "dashed") +
    #geom_text(aes(x= mean_error_rate/2, label=paste0("Mean\n", mean_error_rate), y=50))
    geom_text(aes(label=paste0("mean = ", mean_error_rate), y=Inf, x= mean_error_rate/2), size = 4, vjust = 1)

}
```

```{r function_figure_panels}
all_old_plots <- function(flagstat_path_aa, flagstat_path_nnc, flagstat_path_nc,
                          sys_glob_stat_nnc, sys_glob_stat_nc,
                          ksize, nbhd_reads, database, alphabet){
  aa_paladin_vs_ref <- plot_flagstat_aa(flagstat_path = flagstat_path_aa, 
                                        ksize = ksize, 
                                        nbhd_reads = nbhd_reads, 
                                        database = database, 
                                        alphabet = alphabet) 
  
  nnc_bwa_vs_ref <- plot_flagstat_nnc(flagstat_path = flagstat_path_nnc, 
                                      ksize = ksize,
                                      database = database, 
                                      alphabet = alphabet)
  
  
  nc_bwa_vs_ref <- plot_flagstat_nc(flagstat_path = flagstat_path_nc, 
                                    ksize = ksize, 
                                    database = database, 
                                    alphabet = alphabet)
  
  nnc_bwa_vs_ref_stat <- plot_stat_nnc(sys_glob_stat = sys_glob_stat_nnc,
                                       ksize = ksize, 
                                       database = database, 
                                       alphabet = alphabet)
  
  nc_bwa_vs_ref_stat <- plot_stat_nnc(sys_glob_stat = sys_glob_stat_nc,
                                     ksize = ksize, 
                                     database = database, 
                                     alphabet = alphabet)
  
  ggarrange(aa_paladin_vs_ref, 
            ggarrange(nc_bwa_vs_ref, nnc_bwa_vs_ref, 
                      nc_bwa_vs_ref_stat, nnc_bwa_vs_ref_stat, ncol = 2, nrow = 2), 
            nrow = 2, heights = c(1, 2))
}

all_new_plots <- function(orpheum_json, orpheum_trans, flagstat_path_aa, flagstat_path_nnc, 
                          ksize, alphabet, database, noncoding_estimate){
  frames <- plot_orpheum_translation_frames(orpheum_trans, ksize, alphabet, database)
  aa <- plot_predicted_aa_and_flagstat_aa(orpheum_json, flagstat_path_aa, ksize, alphabet, database, noncoding_estimate)
  nnc <- plot_predicted_nnc_and_flagstat_nnc(orpheum_json, flagstat_path_nnc, ksize, alphabet, database, noncoding_estimate)
  ggarrange(frames, aa, nnc, ncol = 2, nrow = 2)
}
```

## Background

+ The average bacterial genome is 5 Mbp and encodes 5000 proteins ([Land et al. 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4361730/)). As of January 2014:
    + The largest bacterial genome in GenBank was *Sorangium cellulosum* strain So0157-2, with 14,782,125 bp encoding 11,599 genes
    + The smallest bacterial genome in GenBank was *Candidatus Nasuia deltocephalinicola* strain NAS-ALF with 112,091 bp encoding 137 genes.
+ It is estimated that 88% (40-97%) of the bacterial genome is protein coding ([Land et al. 2015](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4361730/)).
+ It is not currently known how many alternate/overlapping ORFs exist in a bacterial genome.
    + Computational estimates suggest there are 10-400 alternate ORFs per genome ([Ardern et al. 2020](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7457138/)).
    + Experimental evidence suggest ~100 alternate overlapping ORFs per genome ([Zehentner et al. 2020](https://www.biorxiv.org/content/10.1101/2020.11.18.388249v1.full.pdf))
+ Using 76 isolates in RefSeq, on average 89.2% of the *R. gnavus* genome is coding. While we don't know if CDSs will follow the same percentages in metagenome assembly graph query neighborhoods, this is a good approximate estimate to aim for for recovery of coding sequences.

## Data description

The 605 gut microbiome metagenomes analyzed in this repository were originally analysed in the [2020-ibd](https://github.com/dib-lab/2020-ibd) repository as a meta-cohort of IBD subtypes (CD, UC, and nonIBD).
Assembly graph neighborhoods for the query genome *R. gnavus* were extracted with spacegraphcats.

## Summary tables

```{r read_and_format_orpheum_json}
orph_plass_k7 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/plass_assembly/protein_ksize7/*json",
                                              database = "plass_assembly",
                                              alphabet = "protein",
                                              ksize = 7)

orph_plass_k10 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/plass_assembly/protein_ksize10/*json",
                                               database = "plass_assembly",
                                               alphabet = "protein",
                                               ksize = 10)

orph_plass_k15 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/plass_assembly/dayhoff_ksize15/*json",
                                              database = "plass_assembly",
                                              alphabet = "dayhoff",
                                              ksize = 15)

orph_plass_k17 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/plass_assembly/dayhoff_ksize17/*json",
                                               database = "plass_assembly",
                                               alphabet = "dayhoff",
                                               ksize = 17)

orph_roary_k6 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/protein_ksize6/*json",
                                              database = "roary_with_megahit_and_isolates",
                                              alphabet = "protein",
                                              ksize = 6)

orph_roary_k7 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/protein_ksize7/*json",
                                              database = "roary_with_megahit_and_isolates",
                                              alphabet = "protein",
                                              ksize = 7)

orph_roary_k10 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/protein_ksize10/*json",
                                               database = "roary_with_megahit_and_isolates",
                                               alphabet = "protein",
                                               ksize = 10)

orph_roary_k11 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize11/*json",
                                               database = "roary_with_megahit_and_isolates",
                                               alphabet = "dayhoff",
                                               ksize = 11)

orph_roary_k13 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize13/*json",
                                               database = "roary_with_megahit_and_isolates",
                                               alphabet = "dayhoff",
                                               ksize = 13)

orph_roary_k15 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize15/*json",
                                               database = "roary_with_megahit_and_isolates",
                                               alphabet = "dayhoff",
                                               ksize = 15)

orph_roary_k17 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize17/*json",
                                               database = "roary_with_megahit_and_isolates",
                                               alphabet = "dayhoff",
                                               ksize = 17)

orph_ruminococcusB_k6 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/ruminococcusB/protein_ksize6/*json",
                                              database = "ruminococcusB",
                                              alphabet = "protein",
                                              ksize = 6)

orph_ruminococcusB_k7 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/ruminococcusB/protein_ksize7/*json",
                                              database = "ruminococcusB",
                                              alphabet = "protein",
                                              ksize = 7)

orph_ruminococcusB_k10 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/ruminococcusB/protein_ksize10/*json",
                                              database = "ruminococcusB",
                                              alphabet = "protein",
                                              ksize = 10)

orph_f__Lachnospiraceae_k6 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/f__Lachnospiraceae/protein_ksize6/*json",
                                              database = "f__Lachnospiraceae",
                                              alphabet = "protein",
                                              ksize = 6)

orph_f__Lachnospiraceae_k7 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/f__Lachnospiraceae/protein_ksize7/*json",
                                              database = "f__Lachnospiraceae",
                                              alphabet = "protein",
                                              ksize = 7)

orph_f__Lachnospiraceae_k10 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/f__Lachnospiraceae/protein_ksize10/*json",
                                              database = "f__Lachnospiraceae",
                                              alphabet = "protein",
                                              ksize = 10)

orph_p__Firmicutes_A_k7 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/p__Firmicutes_A/protein_ksize7/*json",
                                              database = "p__Firmicutes_A",
                                              alphabet = "protein",
                                              ksize = 7)

orph_p__Firmicutes_A_k10 <- read_and_format_orpheum_json(sys_glob_path = "outputs/orpheum/p__Firmicutes_A/protein_ksize10/*json",
                                              database = "p__Firmicutes_A",
                                              alphabet = "protein",
                                              ksize = 10)

all_orph <- bind_rows(orph_plass_k7, orph_plass_k10, orph_plass_k15, orph_plass_k17,
                      orph_roary_k6, orph_roary_k7, orph_roary_k10, orph_roary_k11,
                      orph_roary_k13, orph_roary_k15, orph_roary_k17, orph_ruminococcusB_k6,
                      orph_ruminococcusB_k7, orph_f__Lachnospiraceae_k6, orph_f__Lachnospiraceae_k7,
                      orph_f__Lachnospiraceae_k10, orph_p__Firmicutes_A_k7, orph_p__Firmicutes_A_k10)

all_orph_summary <- all_orph %>%
  group_by(database, alphabet, ksize) %>%
  summarise(across(starts_with(c('t', 'cod', 'non', 'low', 'read', 'nbhd')), mean)) %>%
  select(database, alphabet, ksize, nbhd_reads, coding, non_coding,
         translation_is_shorter_than_peptide_k_mer_size_1, 
         translation_frame_has_stop_codon_s)

kable(all_orph_summary, digits = 0, booktabs = T,
      col.names = c("database", "alphabet", "ksize", "reads", "coding", "non-coding", "too short", "stop codon"))
```

```{r orpheum_as_pct}
all_orph_summary_pct <- all_orph %>%
  group_by(database, alphabet, ksize) %>%
  summarise(across(starts_with(c('t', 'cod', 'non', 'low', 'read', 'nbhd')), mean)) %>%
  mutate(coding = (coding/nbhd_reads) * 100,
         non_coding = (non_coding/nbhd_reads) * 100,
         translation_is_shorter_than_peptide_k_mer_size_1 = (translation_is_shorter_than_peptide_k_mer_size_1/nbhd_reads) *100,
         translation_frame_has_stop_codon_s = (translation_frame_has_stop_codon_s/nbhd_reads)*100) %>%
  select(database, ksize, coding, non_coding,
         translation_is_shorter_than_peptide_k_mer_size_1, 
         translation_frame_has_stop_codon_s)


kable(all_orph_summary_pct, digits = 0, booktabs = T,
      col.names = c("database", "alphabet", "ksize", "coding", "non-coding", "too short", "stop codon"))
```

```{r read_and_format_orpheum_json_translation}
orph_trans_plass_k7 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/plass_assembly/protein_ksize7/*json",
                                                                database = "plass_assembly",
                                                                alphabet = "protein",
                                                                ksize = 7)

orph_trans_plass_k10 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/plass_assembly/protein_ksize10/*json",
                                                                 database = "plass_assembly",
                                                                 alphabet = "protein",
                                                                 ksize = 10)

orph_trans_plass_k15 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/plass_assembly/dayhoff_ksize15/*json",
                                                                 database = "plass_assembly",
                                                                 alphabet = "dayhoff",
                                                                 ksize = 15)

orph_trans_plass_k17 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/plass_assembly/dayhoff_ksize17/*json",
                                                                 database = "plass_assembly",
                                                                 alphabet = "dayhoff",
                                                                 ksize = 17)

orph_trans_roary_k6 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/protein_ksize6/*json",
                                                                database = "roary",
                                                                alphabet = "protein",
                                                                ksize = 6)

orph_trans_roary_k7 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/protein_ksize7/*json",
                                                                database = "roary",
                                                                alphabet = "protein",
                                                                ksize = 7)

orph_trans_roary_k10 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/protein_ksize10/*json",
                                                                 database = "roary",
                                                                 alphabet = "protein",
                                                                 ksize = 10)

orph_trans_roary_k11 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize11/*json",
                                                                 database = "roary",
                                                                 alphabet = "dayhoff",
                                                                 ksize = 11)

orph_trans_roary_k13 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize13/*json",
                                                                 database = "roary",
                                                                 alphabet = "dayhoff",
                                                                 ksize = 13)

orph_trans_roary_k15 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize15/*json",
                                                                 database = "roary",
                                                                 alphabet = "dayhoff",
                                                                 ksize = 15)

orph_trans_roary_k17 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/roary_with_megahit_and_isolates/dayhoff_ksize17/*json",
                                                                 database = "roary",
                                                                 alphabet = "dayhoff",
                                                                 ksize = 17)

orph_trans_ruminococcusB_k6 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/ruminococcusB/protein_ksize6/*json",
                                                                        database = "ruminococcusB",
                                                                        alphabet = "protein",
                                                                        ksize = 6)

orph_trans_ruminococcusB_k7 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/ruminococcusB/protein_ksize7/*json",
                                                                        database = "ruminococcusB",
                                                                        alphabet = "protein",
                                                                        ksize = 7)

orph_trans_ruminococcusB_k10 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/ruminococcusB/protein_ksize10/*json",
                                                                         database = "ruminococcusB",
                                                                         alphabet = "protein",
                                                                         ksize = 10)

orph_trans_f__Lachnospiraceae_k6 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/f__Lachnospiraceae/protein_ksize6/*json",
                                                                             database = "f__Lachnospiraceae",
                                                                             alphabet = "protein",
                                                                             ksize = 6)

orph_trans_f__Lachnospiraceae_k7 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/f__Lachnospiraceae/protein_ksize7/*json",
                                                                             database = "f__Lachnospiraceae",
                                                                             alphabet = "protein",
                                                                             ksize = 7)

orph_trans_f__Lachnospiraceae_k10 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/f__Lachnospiraceae/protein_ksize10/*json",
                                                                              database = "f__Lachnospiraceae",
                                                                              alphabet = "protein",
                                                                              ksize = 10)

orph_trans_p__Firmicutes_A_k7 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/p__Firmicutes_A/protein_ksize7/*json",
                                                                          database = "p__Firmicutes_A",
                                                                          alphabet = "protein",
                                                                          ksize = 7)

orph_trans_p__Firmicutes_A_k10 <- read_and_format_orpheum_json_translation(sys_glob_path = "outputs/orpheum/p__Firmicutes_A/protein_ksize10/*json",
                                                                           database = "p__Firmicutes_A",
                                                                           alphabet = "protein",
                                                                           ksize = 10)



```

## Controls

```{r nbhd_reads}
nbhd_reads <- orph_plass_k10 %>%
  select(sample, nbhd_reads)
```

```{r control, fig.cap="\\label{fig:control}Mean percent mapped reads against AA reference pangenome. On average, **66.7% of reads** mapped against the pagenome reference using the paladin amino acid mapper. This number estimates the lower limit of reads that should be protein coding; at least this many reads should.", fig.height=2, fig.width = 4}
nuc_reads_vs_aa_ref <- read_tsv("outputs/rgnv_sgc_original_paladin/multiqc_data/multiqc_samtools_flagstat.txt") %>%
  mutate(sample = gsub("_GCF_900036035.1_RGNV35913_genomic.fna", "", Sample)) %>%
  left_join(nbhd_reads, by = "sample") %>%
  select(sample, nbhd_reads, mapped_passed) %>%
  mutate(mapped_passed_pct = (mapped_passed/nbhd_reads)*100)

mean_mapped_passed_pct <- nuc_reads_vs_aa_ref %>% 
  pull(mapped_passed_pct) %>% 
  mean() %>%
  signif(3)

control_plt <- ggplot(nuc_reads_vs_aa_ref, aes(x = mapped_passed_pct)) +
  geom_density() +
  xlim(0, 100) +
  geom_vline(aes(xintercept = mean(mapped_passed_pct)), col='grey', linetype = "dashed") +
  geom_text(aes(x=mean_mapped_passed_pct/2, label=paste0("Mean\n", mean_mapped_passed_pct), y=.02)) +
  ggplot_base 
control_plt
```

## Experimental DB1: PLASS assembly


```{r plass_plots, fig.height = 5}
all_new_plots(orpheum_json = orph_plass_k7, 
              orpheum_trans = orph_trans_plass_k7,
              flagstat_path_aa = "outputs/aa_paladin//plass_assembly/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/plass_assembly/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 7, 
              alphabet = "protein",
              database = "plass_assembly", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_plass_k10, 
              orpheum_trans = orph_trans_plass_k10,
              flagstat_path_aa = "outputs/aa_paladin/plass_assembly/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/plass_assembly/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 10, 
              alphabet = "protein",
              database = "plass_assembly", 
              noncoding_estimate = 10.8)
```

## Experimental DB2: Reference pangenome

+ The <u>reference pangenome</u> is ~37k prokka predicted protein sequences from 76 *R. gnavus* isolates from RefSeq and megahit assemblies of R. gnavus sgc nbhds


```{r roary_plots, fig.height = 5}

all_new_plots(orpheum_json = orph_roary_k6, 
              orpheum_trans = orph_trans_roary_k6,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/protein_ksize6/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/protein_ksize6/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 6, 
              alphabet = "protein",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_roary_k7, 
              orpheum_trans = orph_trans_roary_k7,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 7, 
              alphabet = "protein",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_roary_k10, 
              orpheum_trans = orph_trans_roary_k10,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 10, 
              alphabet = "protein",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_roary_k11, 
              orpheum_trans = orph_trans_roary_k11,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/dayhoff_ksize11/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/dayhoff_ksize11/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 11, 
              alphabet = "dayhoff",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_roary_k13, 
              orpheum_trans = orph_trans_roary_k13,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/dayhoff_ksize13/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/dayhoff_ksize13/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 13, 
              alphabet = "dayhoff",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_roary_k15, 
              orpheum_trans = orph_trans_roary_k15,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/dayhoff_ksize15/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/dayhoff_ksize15/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 15, 
              alphabet = "dayhoff",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_roary_k17, 
              orpheum_trans = orph_trans_roary_k17,
              flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/dayhoff_ksize17/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/dayhoff_ksize17/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 17, 
              alphabet = "dayhoff",
              database = "roary_with_megahit_and_isolates", 
              noncoding_estimate = 10.8)


```

## Experimental DB3: GTDB ruminococcusB

```{r ruminococcusB_plots, fig.height = 5}
all_new_plots(orpheum_json = orph_ruminococcusB_k6, 
              orpheum_trans = orph_trans_ruminococcusB_k6,
              flagstat_path_aa = "outputs/aa_paladin/ruminococcusB/protein_ksize6/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/ruminococcusB/protein_ksize6/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 6, 
              alphabet = "protein",
              database = "ruminococcusB", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_ruminococcusB_k7, 
              orpheum_trans = orph_trans_ruminococcusB_k7,
              flagstat_path_aa = "outputs/aa_paladin/ruminococcusB/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/ruminococcusB/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 7, 
              alphabet = "protein",
              database = "ruminococcusB", 
              noncoding_estimate = 10.8)
```

## Experimental DB4: GTDB family f__Lachnospiraceae

```{r f__Lachnospiraceae_plots, fig.height = 5}
all_new_plots(orpheum_json = orph_f__Lachnospiraceae_k6, 
              orpheum_trans = orph_trans_f__Lachnospiraceae_k6,
              flagstat_path_aa = "outputs/aa_paladin/f__Lachnospiraceae/protein_ksize6/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/f__Lachnospiraceae/protein_ksize6/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 6, 
              alphabet = "protein",
              database = "f__Lachnospiraceae", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_f__Lachnospiraceae_k7, 
              orpheum_trans = orph_trans_f__Lachnospiraceae_k7,
              flagstat_path_aa = "outputs/aa_paladin/f__Lachnospiraceae/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/f__Lachnospiraceae/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 7, 
              alphabet = "protein",
              database = "f__Lachnospiraceae", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_f__Lachnospiraceae_k10, 
              orpheum_trans = orph_trans_f__Lachnospiraceae_k10,
              flagstat_path_aa = "outputs/aa_paladin/f__Lachnospiraceae/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/f__Lachnospiraceae/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 10, 
              alphabet = "protein",
              database = "f__Lachnospiraceae", 
              noncoding_estimate = 10.8)
```

## Experimental DB5: GTDB family p__Firmicutes_A

```{r p__Firmicutes_A_plots, fig.height = 5}
all_new_plots(orpheum_json = orph_p__Firmicutes_A_k7, 
              orpheum_trans = orph_trans_p__Firmicutes_A_k7,
              flagstat_path_aa = "outputs/aa_paladin/p__Firmicutes_A/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/p__Firmicutes_A/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 7, 
              alphabet = "protein",
              database = "p__Firmicutes_A", 
              noncoding_estimate = 10.8)

all_new_plots(orpheum_json = orph_p__Firmicutes_A_k10, 
              orpheum_trans = orph_trans_p__Firmicutes_A_k10,
              flagstat_path_aa = "outputs/aa_paladin/p__Firmicutes_A/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc = "outputs/nuc_noncoding_bwa/p__Firmicutes_A/protein_ksize10/multiqc_data/multiqc_samtools_flagstat.txt",
              ksize = 10,
              alphabet = "protein",
              database = "p__Firmicutes_A", 
              noncoding_estimate = 10.8)
```

## Conclusions
 
+ PLASS is too promiscuous of an assembler to use to generate the reference DB; it leads to too many off-ORF predictions for reads.
+ The reference pangenome seems to work pretty well, it marginally outperforms just mapping reads with paladin, and all the coding reads map to the CDS in the pangenome.

## TODO

+ GTDB reps/GTDB as db
+ Search for shine dalgarno sequences in non-coding reads (consensus seq AGGAGG). Although SD seqs are only 8 bp ahead of AUG (start codon), so may not be in majorit of seq. May be worth looking into other conserved intergenic seqs. 

## References

+ Land et al. 2015: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4361730/
+ Ardern et al. 2020: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7457138/
+ Zehentner et al. 2020: https://www.biorxiv.org/content/10.1101/2020.11.18.388249v1.full.pdf

## Supplementary figure panels

```{r old_panel_of_figures}
all_old_plots(flagstat_path_aa = "outputs/aa_paladin/roary_with_megahit_and_isolates/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nnc ="outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt", 
              flagstat_path_nc = "outputs/nuc_coding_bwa/roary_with_megahit_and_isolates/protein_ksize7/multiqc_data/multiqc_samtools_flagstat.txt",
              sys_glob_stat_nnc = "outputs/nuc_noncoding_bwa/roary_with_megahit_and_isolates/protein_ksize7/*.stat", 
              sys_glob_stat_nc = "outputs/nuc_coding_bwa/roary_with_megahit_and_isolates/protein_ksize7/*.stat",
              ksize = 7, nbhd_reads = nbhd_reads, 
              database = "roary", alphabet = "protein")
```

